tests/baselines/reference/inferTypePredicates.types

```diff
@@ -0,0 +1,932 @@
+
+  return typeof x === 'number';
+>typeof x === 'number' : boolean
+>typeof x : "string" | "number" | "bigint" | "boolean" | "symbol" | "undefined" | "object" | "function"
+>x : number | null
+>'number' : "number"
+}
+
+// The type predicate must remain valid when the function is called with subtypes.
+function isShortString(x: unknown) {
+>isShortString : (x: unknown) => boolean
+>x : unknown
+
+  return typeof x === "string" && x.length < 10;
+>typeof x === "string" && x.length < 10 : boolean
+>typeof x === "string" : boolean
+>typeof x : "string" | "number" | "bigint" | "boolean" | "symbol" | "undefined" | "object" | "function"
+>x : unknown
+>"string" : "string"
+>x.length < 10 : boolean
+>x.length : number
+>x : string
+>length : number
+>10 : 10
+}
+
+declare let str: string;
+>str : string
+
+if (isShortString(str)) {
+>isShortString(str) : boolean
+>isShortString : (x: unknown) => boolean
+>str : string
+
+  str.charAt(0);  // should ok
+>str.charAt(0) : string
+>str.charAt : (pos: number) => string
+>str : string
+>charAt : (pos: number) => string
+>0 : 0
+
+} else {
+  str.charAt(0);  // should ok
+>str.charAt(0) : string
+>str.charAt : (pos: number) => string
+>str : string
+>charAt : (pos: number) => string
+>0 : 0
+}
+
+function isStringFromUnknown(x: unknown) {
+>isStringFromUnknown : (x: unknown) => x is string
+>x : unknown
+
+  return typeof x === "string";
+>typeof x === "string" : boolean
+>typeof x : "string" | "number" | "bigint" | "boolean" | "symbol" | "undefined" | "object" | "function"
+>x : unknown
+>"string" : "string"
+}
+if (isStringFromUnknown(str)) {
+>isStringFromUnknown(str) : boolean
+>isStringFromUnknown : (x: unknown) => x is string
+>str : string
+
+  str.charAt(0);  // should OK
+>str.charAt(0) : string
+>str.charAt : (pos: number) => string
+>str : string
+>charAt : (pos: number) => string
+>0 : 0
+
+} else {
+  let t: never = str;  // should OK
+>t : never
+>str : never
+}
+
+// infer a union type
+function isNumOrStr(x: unknown) {
+>isNumOrStr : (x: unknown) => x is string | number
+>x : unknown
+
+  return (typeof x === "number" || typeof x === "string");
+>(typeof x === "number" || typeof x === "string") : boolean
+>typeof x === "number" || typeof x === "string" : boolean
+>typeof x === "number" : boolean
+>typeof x : "string" | "number" | "bigint" | "boolean" | "symbol" | "undefined" | "object" | "function"
+>x : unknown
+>"number" : "number"
+>typeof x === "string" : boolean
+>typeof x : "string" | "number" | "bigint" | "boolean" | "symbol" | "undefined" | "object" | "function"
+>x : unknown
+>"string" : "string"
+}
+declare let unk: unknown;
+>unk : unknown
+
+if (isNumOrStr(unk)) {
+>isNumOrStr(unk) : boolean
+>isNumOrStr : (x: unknown) => x is string | number
+>unk : unknown
+
+  let t: number | string = unk;  // should ok
+>t : string | number
+>unk : string | number
+}
+
+// A function can be a type predicate even if it throws.
+function assertAndPredicate(x: string | number | Date) {

```

Comment:
@ahejlsberg I just wanted to note that we infer a type predicate for this function with your change whereas we didn't before (that was the reason for the third call to `getFlowTypeOfReference`). This type guard isn't wrong, per se, just imprecise because if this function returns `false` then `x` is a `string`, not `string | Date`. I doubt this pattern is very common in the wild.

Reply:
I think the new behavior is preferable. We're not reflecting the fact that the function never returns for certain input types, but that shouldn't keep us from more accurately reflecting the behavior when the function actually returns. I mean, it's even more wrong to infer the function has no effect at all.